src/cargo/ops/cargo_package/vcs.rs

```diff
@@ -206,6 +209,36 @@ fn dirty_metadata_paths(pkg: &Package, repo: &git2::Repository) -> CargoResult<V
     Ok(dirty_files)
 }
 
+/// Checks whether source files are symlinks and have been modified.
+///
+/// This is required because those paths may link to a file outside the
+/// current package root, but still under the git workdir, affecting the
+/// final packaged `.crate` file.
+fn dirty_symlinks(
+    pkg: &Package,
+    repo: &git2::Repository,
+    src_files: &[PathEntry],
+) -> CargoResult<HashSet<PathBuf>> {
+    let workdir = repo.workdir().unwrap();
+    let mut dirty_symlinks = HashSet::new();
+    for rel_path in src_files
+        .iter()
+        .filter(|p| p.is_symlink_or_under_symlink())
+        .map(|p| p.as_ref().as_path())
+        // If inside package root. Don't bother checking git status.
+        .filter(|p| paths::strip_prefix_canonical(*p, pkg.root()).is_err())
+        // Handle files outside package root but under git workdir,
+        .filter_map(|p| paths::strip_prefix_canonical(p, workdir).ok())
+    {
+        // TODO: Should we warn users if there are like thousands of symlinks,
+        // which may hurt the performance?
```

Comment:
If we need to track this, we should probably do this in a way that has visibility

Reply:
Did you mean opening an issue?

I can do a benchmark and find a reasonable number to emit a warning.

Reply:
Alternatively, if this is speculation, should we wait on a user to have a problem?

Reply:
Sounds like we can just remove the TODO comment.